### PR TITLE
add Nordic JEDEC QSPI NOR flash driver

### PR DESCRIPTION
**Pull request by kamlaz**
Add QSPI flash driver support for Nordic nRF52840 chip (PCA10056 development board).

### CODE REVIEW

File: drivers/flash/Kconfig.nordic_qspi_nor
Comment (ulfalizer) [inline drivers/flash/Kconfig.nordic_qspi_nor]: Add a blank line after the `if` to make it balanced with the `endif`. See https://docs.zephyrproject.org/latest/guides/kconfig/tips.html#header-comments-and-other-nits.

File: drivers/flash/Kconfig.nordic_qspi_nor
Comment (ulfalizer) [inline drivers/flash/Kconfig.nordic_qspi_nor]: Accidental tab instead of space.

File: dts/bindings/flash_controller/nordic,nrf-qspi.yaml
Comment (ulfalizer) [inline dts/bindings/flash_controller/nordic,nrf-qspi.yaml]: Missing newline.

File: dts/bindings/mtd/nordic,qspi-nor.yaml
Comment (ulfalizer) [inline dts/bindings/mtd/nordic,qspi-nor.yaml]: This will show up in the header as a comment above the generated macros, so might be better to make it "QSPI NOR flash supporting the JEDEC CFI interface".

**Review by ulfalizer — COMMENTED** [review]
Some nits.

File: samples/drivers/spi_flash/boards/nrf52840_pca10056.conf
Comment (nvlsianpu) [inline samples/drivers/spi_flash/boards/nrf52840_pca10056.conf]: rather add copyright, not replace

File: drivers/flash/nrf_qspi_nor.c
Comment (nvlsianpu) [inline drivers/flash/nrf_qspi_nor.c]: @brief, let's keep doxygen comment consistent

File: drivers/flash/nrf_qspi_nor.c
Comment (nvlsianpu) [inline drivers/flash/nrf_qspi_nor.c]: to many blank lines

File: drivers/flash/nrf_qspi_nor.c
Comment (nvlsianpu) [inline drivers/flash/nrf_qspi_nor.c]: minor suggestion: LUT maybe?
```
const int ret_array[] = {NRF_QSPI_READOC_FASTREAD, NRF_QSPI_READOC_READ2O, NRF_QSPI_READOC_READ2IO, NRF_QSPI_READOC_READ4O, NRF_QSPI_READOC_READ4IO};

if (lines < ARRAY_SIZE( ret_array[])) {
  ret = ret_array[lines];
}
```

above as well

File: drivers/flash/nrf_qspi_nor.c
Comment (nvlsianpu) [inline drivers/flash/nrf_qspi_nor.c]: remove dash.
in other places of this file is saw colon after a perimeter line, I prefer consistency like `@param param_name what_param_does`

File: drivers/flash/nrf_qspi_nor.c
Comment (nvlsianpu) [inline drivers/flash/nrf_qspi_nor.c]: to many white space

File: drivers/flash/nrf_qspi_nor.c
Comment (nvlsianpu) [inline drivers/flash/nrf_qspi_nor.c]:
```suggestion
	.size = INST_0_BYTES
```

File: drivers/flash/nrf_qspi_nor.c
Comment (nvlsianpu) [inline drivers/flash/nrf_qspi_nor.c]: move INST_0_BYTES outside of ifdefry in case using in device initialization macro

File: samples/drivers/spi_flash/boards/nrf52840_pca10056.conf
Comment (pabigot) [reply samples/drivers/spi_flash/boards/nrf52840_pca10056.conf]: This commit is actually from me, but authorship got lost during development, and I removed  this because in the churn around this I'd forgotten it was originally added on my own time not for Nordic.  But yes, it probably should have both, just for history.

I would rather that the commit was restored to the original authorship and sign-off, if that can be done.  `git commit --amend -C sha-of-original` should do it.

File: drivers/flash/nrf_qspi_nor.c
Comment (nvlsianpu) [inline drivers/flash/nrf_qspi_nor.c]: Does it make sense to use LUT in the loop searching here?

File: drivers/flash/nrf_qspi_nor.c
Comment (anangl) [inline drivers/flash/nrf_qspi_nor.c]: `res` codes for partial erases are not checked in this loop, only the translated code for the last result is returned.

File: drivers/flash/nrf_qspi_nor.c
Comment (anangl) [inline drivers/flash/nrf_qspi_nor.c]: `addr` is not checked here, also for sizes greater then the size of the chip the loop will try to do partial erases.

File: drivers/flash/nrf_qspi_nor.c
Comment (anangl) [inline drivers/flash/nrf_qspi_nor.c]: In case this is not successful, the function should no attempt to call `qspi_send_cmd`.

File: drivers/flash/nrf_qspi_nor.c
Comment (anangl) [inline drivers/flash/nrf_qspi_nor.c]: This code will get translated in `qspi_get_zephyr_ret_code` to `-EBUSY`. I guess this was not the intent.

File: drivers/flash/nrf_qspi_nor.c
Comment (anangl) [inline drivers/flash/nrf_qspi_nor.c]: This can be `const`.

File: drivers/flash/nrf_qspi_nor.c
Comment (anangl) [inline drivers/flash/nrf_qspi_nor.c]: This can be `const`.

File: drivers/flash/nrf_qspi_nor.c
Comment (anangl) [inline drivers/flash/nrf_qspi_nor.c]: What's the point in involving `rx_buf` while you could use `dest` and `size` directly here?
Same goes for the call to `nrfx_qspi_write` below.

**Review by anangl — DISMISSED** [review]

File: samples/drivers/spi_flash/boards/nrf52840_pca10056.conf
Comment (pabigot) [reply samples/drivers/spi_flash/boards/nrf52840_pca10056.conf]: The board-specific overlay is appended to the project overlay which selects `CONFIG_SPI_NOR`, so the overlay needs to deselect it.

If `CONFIG_SPI_2` remains enabled I believe something else broke, but I don't remember what it was.  Probably to do with the subordinate node on that device, or the flash partitions bound to it.

Comment (anangl) [reply samples/drivers/spi_flash/boards/nrf52840_pca10056.conf]: I don't think I understand. Which project overlay are you referring to? This sample's [prj.conf](https://github.com/zephyrproject-rtos/zephyr/blob/f638a7e637a341a7f3e25ce281cfa6957ab3d879/samples/drivers/spi_flash/prj.conf) does not select `CONFIG_QSPI_NOR`.

Comment (pabigot) [reply samples/drivers/spi_flash/boards/nrf52840_pca10056.conf]: OK, I see; this is a result of making the overlay consistent with the other samples that are spi-nor specific.  

I'd say try removing the two lines and see what happens.

File: drivers/flash/Kconfig.nordic_qspi_nor
Comment (kamlaz) [reply drivers/flash/Kconfig.nordic_qspi_nor]: Good idea. I will change it.

File: drivers/flash/Kconfig.nordic_qspi_nor
Comment (kamlaz) [reply drivers/flash/Kconfig.nordic_qspi_nor]: Yes, it should be changed.

File: drivers/flash/Kconfig.nordic_qspi_nor
Comment (kamlaz) [reply drivers/flash/Kconfig.nordic_qspi_nor]: Right - it could be removed.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: That is right - I missed that.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Fixed

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Exactly - silly mistake.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Added `const`

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Added `const`

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Indeed, it is not needed anymore. Removed it.

File: samples/drivers/spi_flash/boards/nrf52840_pca10056.conf
Comment (kamlaz) [reply samples/drivers/spi_flash/boards/nrf52840_pca10056.conf]: Removing `CONFIG_SPI_NOR` has no effect, but removing `CONFIG_SPI_2` causes that application is unable to find driver using `device_get_binding`.  Perhaps it is due to pins assigned to that `SPI` instance (CS pins) are also used by QSPI.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: I have added check condition.

File: drivers/flash/Kconfig.nordic_qspi_nor
Comment (anangl) [inline drivers/flash/Kconfig.nordic_qspi_nor]: typo - Register

File: samples/drivers/spi_flash/boards/nrf52840_pca10056.conf
Comment (anangl) [reply samples/drivers/spi_flash/boards/nrf52840_pca10056.conf]: So the problem is invalid default configuration of the board (same pins assigned to `spi2` and `qspi`, and both these peripherals enabled). And that is what should be corrected. By disabling `SPI_2` in this application you are only masking the problem.

File: drivers/flash/nrf_qspi_nor.c
Comment (anangl) [inline drivers/flash/nrf_qspi_nor.c]: Errors should be rather reported with `LOG_ERR`.

File: drivers/flash/Kconfig.nordic_qspi_nor
Comment (kamlaz) [reply drivers/flash/Kconfig.nordic_qspi_nor]: Fixed

File: drivers/flash/Kconfig.nordic_qspi_nor
Comment (kamlaz) [reply drivers/flash/Kconfig.nordic_qspi_nor]: Changed prefixes.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Changed.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Fixed.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Indeed. Fixed.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Added `static`

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Changed `LOG_DBG` to `LOG_ERR`.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: I have added `LOG_ERR` inside the `qspi_nor_read_id` to provide info about extected ID an obtained ID.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Added same address check algorithm as in soc_flash_nrf.c

**Comment by carlescufi** [comment]
@anangl @pabigot please re-review.

File: drivers/flash/nrf_qspi_nor.c
Comment (nvlsianpu) [inline drivers/flash/nrf_qspi_nor.c]: this comparison stages look erroneous as ensure else case in second compassion.

**Review by nvlsianpu — CHANGES_REQUESTED** [review]

File: drivers/flash/nrf_qspi_nor.c
Comment (nvlsianpu) [reply drivers/flash/nrf_qspi_nor.c]: just 2000000ul

Comment (pabigot) [reply drivers/flash/nrf_qspi_nor.c]: IMO this should be `2000000UL`, i.e. upper case.  We don't want a lower case `l` because it looks too much like `1`, so for consistency it should be `UL`.  Also see line 172 where `32000000uL` looks weird.

Not a blocking comment, just a preference.

**Review by pabigot — COMMENTED** [review]
Just one comment for now.  I'll approve when I rebase my littlefs test support on this and verify the behavior again.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: I have added changes due to yours request.

**Review by pabigot — CHANGES_REQUESTED** [review]
While testing I noticed that `samples/subys/fs/littlefs` for nrf52840_pca10056 has a configuration override that targets the external flash memory, but did not include a devicetree overlay that redefined the partition table so the storage partition is on that device.

Please look at https://github.com/pabigot/zephyr/commits/pr/21535 and cherry-pick https://github.com/pabigot/zephyr/commit/f867c8782354cb48d50f08785a2c5b641b12882a into your PR before the other two commits.  Then your changes to switch from SPI to QSPI for this sample will actually do something.

Alternatively I can submit that fix separately, we can wait until it's merged, then merge this PR.

**Review by pabigot — APPROVED** [review]

**Comment by carlescufi** [comment]
@nvlsianpu please re-review

File: drivers/flash/nrf_qspi_nor.c
Comment (anangl) [reply drivers/flash/nrf_qspi_nor.c]: It seems that you did, but only in `qspi_erase`. Was that the intent?

File: drivers/flash/nrf_qspi_nor.c
Comment (anangl) [inline drivers/flash/nrf_qspi_nor.c]: This is also checked in `qspi_erase` in a bit more extensive way, why do it twice?
And `-EINVAL` is returned there, as well as in `qspi_nor_read` and `qspi_nor_write`, which seems more suitable.

File: drivers/flash/nrf_qspi_nor.c
Comment (anangl) [inline drivers/flash/nrf_qspi_nor.c]: It can be `>=`.

File: drivers/flash/nrf_qspi_nor.c
Comment (nvlsianpu) [reply drivers/flash/nrf_qspi_nor.c]: @kamlaz please apply hat was requested here - it's the bug

**Review by nvlsianpu — COMMENTED** [review]
Possible overflows of `addr + size` is not handled properly for `write` operation.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: I changed that also in write and read commands. I just missed them before.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Right. I reduced this check to one and moved it to the `qspi_nor_erase`.

File: drivers/flash/nrf_qspi_nor.c
Comment (kamlaz) [reply drivers/flash/nrf_qspi_nor.c]: Fixed

**Review by nvlsianpu — APPROVED** [review]

File: drivers/flash/Kconfig.nordic_qspi_nor
Comment (anangl) [inline drivers/flash/Kconfig.nordic_qspi_nor]: This is not true, this device does not use an SPI driver.

File: drivers/flash/Kconfig.nordic_qspi_nor
Comment (anangl) [inline drivers/flash/Kconfig.nordic_qspi_nor]: As this option has no prompt, users will not be able to modify it in menuconfig or in an application's `prj.conf`. Was that the intent? I see that in `Kconfig.nor` the situation is the same, but I'm actually not sure if that was intended there either.

File: drivers/flash/Kconfig.nordic_qspi_nor
Comment (kamlaz) [reply drivers/flash/Kconfig.nordic_qspi_nor]: I removed misleading part.

File: drivers/flash/Kconfig.nordic_qspi_nor
Comment (kamlaz) [reply drivers/flash/Kconfig.nordic_qspi_nor]: I did it intentionally to provide consistency with `Kconfig.nor`. It could be changed, but I think both changes (in `Kconfig.nordic_qspi_nor` and `Kconfig.nor`) schould be done in one commit in another PR (since this one does not touch `spi_nor`).

Comment (pabigot) [reply drivers/flash/Kconfig.nordic_qspi_nor]: I agree with that.  I doubt it was intentional in either case.  I can change it in both later, or we could make progress on https://github.com/zephyrproject-rtos/zephyr/issues/20348#issuecomment-564108715 which could lead to automatic management of initialization order based on dependencies.

**Review by anangl — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
